### ISSUE TITLE
bug: spurious fuzzy search result items

### ISSUE THREAD

**Issue by huaixv** [nvim-neo-tree/neo-tree.nvim#836]
#721 implements the new fuzzy search algorithm.
The current implementation gives spurious search results as it directly uses `glob or regex` as search pattern, **without** prepending it with the `path` to be searched.

https://github.com/nvim-neo-tree/neo-tree.nvim/blob/20c2f2f5ba083bbb1e37b8bc3d590621434f31e9/lua/neo-tree/sources/filesystem/lib/filter_external.lua#L165

The results will be further fed into `git` to check exclusions, thus creating much more `git` invocations for large projects (in deep path), even causing `EMFILE 24 Too many open files` error, freezing the gui, rendering this feature unusable.

Just prepend `path` to the search pattern will fix it in many cases, but maybe we also need to escape the `path` if `regex` pattern is used. 
```diff
-  append("--", glob or regex or "") 
+  append("--", path .. (glob or regex or ""))
```

**Comment by huaixv** [comment]
> We don't need to add the path to the regex or glob because we are already setting the path as an argument to fd/find.

Yes, pass `path` to fd/find will make them know **where** to search.

But the problem is that the `glob or regex` pattern will be matched against `path .. <relative_filepath>`

Let me give a concrete example, say the search pattern is `dir`,
the implementation will match all files under my PWD `/path/to/my/long/dir/`, not only `/path/to/my/long/dir/dir`,
giving many false positive matches.

**Comment by cseickel** [comment]
Oh I see, you are looking for a search function that will only search one level deep. That's not at all how I or anyone else up until now would expect a fuzzy search to work. Luckily for you it is an option that you can add using the `filesystem.find_args` config:

https://github.com/nvim-neo-tree/neo-tree.nvim/blob/20c2f2f5ba083bbb1e37b8bc3d590621434f31e9/lua/neo-tree/defaults.lua#L423-L428

I think the `fd` command has an option for `--max-depth` that you can set to `1`.

**Comment by huaixv** [comment]
Thank you for your reply.

I found that without `--full-path` option, `fd` works just as expected(not matching pattern with PWD)

Here's another question, why doesn't  `fzy_sort_files` use `opts.find_by_full_path_words` like `find_files` do? (the 5th arg)

https://github.com/nvim-neo-tree/neo-tree.nvim/blob/20c2f2f5ba083bbb1e37b8bc3d590621434f31e9/lua/neo-tree/sources/filesystem/lib/filter_external.lua#L329

https://github.com/nvim-neo-tree/neo-tree.nvim/blob/20c2f2f5ba083bbb1e37b8bc3d590621434f31e9/lua/neo-tree/sources/filesystem/lib/filter_external.lua#L350

**Comment by cseickel** [comment]
> Here's another question, why doesn't `fzy_sort_files` use `opts.find_by_full_path_words` like `find_files` do? (the 5th arg)

It's up to the fzy algorithm to handle filtering and sorting, the external filter is just a first stage so it casts as wide a net as possible and just grabs all files that could possibly satisfy the query.
